### ISSUE TITLE
Restore projects after CFU

### ISSUE THREAD

**Issue by lscheffler** [VFPX/Thor#96]
## 📝 Provide a description of the new feature

Since CFU runs CLOSE all, all projects are gone. What about restoring the config? 
Most simple way is to store the projects to some property of _SCREEN:
```
ADDPROPERTY(_SCREEN,"gaFiles(1,2)")
IF _VFP.PROJECTS.COUNT>0 THEN
 DIMENSION;
  _SCREEN.gaFiles(_VFP.PROJECTS.COUNT,2)
ENDIF &&_VFP.PROJECTS.COUNT>0
_SCREEN.gaFiles = ""

lnProj     = 0
FOR EACH m.loProject IN _VFP.PROJECTS FOXOBJECT
 lnProj							= m.lnProj+1
 _SCREEN.gaFiles(m.lnProj,1)	= m.loProject.NAME
 _SCREEN.gaFiles(m.lnProj,2)	= m.loProject.VISIBLE
ENDFOR &&loProject

CLOSE ALL
************
* run whatever must run
************

IF !EMPTY(_SCREEN.gaFiles(1,1)) THEN
 FOR lnProj = 1 TO ALEN(_SCREEN.gaFiles,1)
  IF _SCREEN.gaFiles(m.lnProj,2) THEN
   MODIFY PROJECT (_SCREEN.gaFiles(m.lnProj,1)) NOWAIT
  ELSE  &&_SCREEN.gaFiles(m.lnProj,2) THEN
   MODIFY PROJECT (_SCREEN.gaFiles(m.lnProj,1)) NOWAIT NOSHOW
  ENDIF &&_SCREEN.gaFiles(m.lnProj,2) THEN
 ENDFOR &&lnProj
ENDIF &&!EMPTY(_SCREEN.gaFiles(1,1)) THEN
REMOVEPROPERTY(_SCREEN,"gaFiles")
```
I think it's not very tricky to include ProjectExplorer into this.

It would be a real help if one uses many projects - and even more while testing Thor plug-ins.  ;)

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

**Comment by Jimrnelson** [comment]
@lscheffler 

Lutz --

I have always been extremely hesitant about making changes to Thor's CFU.  The code is old and fragile and has the danger that if CFU fails, users may not be able to get updates to Thor or any other project.  This happened once and I would rather not relive it.

I am quite curious about the interest in this.  How often are you running CFU?  I run it only once a week, but, to be honest, there are so few updates going on these days that I could run it once a month and not miss much.

Most of your recent work has been in GF6 (as far as I know) and you should be able to test all of your changes to it without ever resorting to CFU.

**Comment by lscheffler** [comment]
Jim

Twice the hour or so right now. Trying to fix my projects to VFPX Deployment. And since the other side is a project open in VFPX Deployment, it's boring to reload.

The problem is more that I have normally a bunch of projects open all the time, and that got lost with CFU. Imagine to reload all this. And the projects are nearly the the first to be loaded, so [this](https://github.com/VFPX/Thor/blob/master/Docs/Thor_running.md) is to late. (In fact, for now my code wraps this call)

I would think this more as a wrapper around CFU. If a project is not closed, MODI PROJ simply does nothing.

**Comment by Jimrnelson** [comment]
Lutz --

This is a easy enough request to implement and will be done in such a way that you can save/restore any other configuration settings you are interested in.  I'll announce it with a Thor News item when it's ready.

I'm still quite curious about how frequently you are running CFU:

> Twice the hour or so right now.

I am trying to guess why that's necessary.  All that CFU does is to unzip a file and run a minor installation program (normally to create a Thor tool or two).   I have never found it necessary to run repeated CFUs on any project.  

So, I'm curious why you find it necessary to do so.

**Comment by rschummer** [comment]
I run CFU once a week, and rarely more than a couple times a week unless someone breaks one of the tools. But I can see how I might use it to update a virtual machine for testing (something I was doing just yesterday for a non-Thor product). Again, rare for my use cases. RickPlease excuse brevity, sent from phoneOn May 21, 2023, at 7:22 AM, Jimrnelson ***@***.***> wrote:﻿
Lutz --
This is a easy enough request to implement and will be done in such a way that you can save/restore any other configuration settings you are interested in.  I'll announce it with a Thor News item when it's ready.
I'm still quite curious about how frequently you are running CFU:

Twice the hour or so right now.

I am trying to guess why that's necessary.  All that CFU does is to unzip a file and run a minor installation program (normally to create a Thor tool or two).   I have never found it necessary to run repeated CFUs on any project.
So, I'm curious why you find it necessary to do so.

**Comment by lscheffler** [comment]
@Jimrnelson Just on an intermediate step to see if something using a Tool installed via Thor is working by an other Tool. I understand how both are running without Thor, but that's not the problem. I'm back to my former how-to-debug-this problem. The projects are non public so far, at least I managed to set the update files local. But still I need to run CFU itself.

**Comment by Jimrnelson** [comment]
@lscheffler 

>  I'm back to my former how-to-debug-this problem. The projects are non public so far, at least I managed to set the update files local. But still I need to run CFU itself.

Please explain in detail what your how-to-debug problem is.   CFU is used to distribute projects to others, but I can't see how it is needed for your own uses.

I believe we should be able to find a way to make this easier for you, especially these days when you are so active.

**Comment by lscheffler** [comment]
@Jimrnelson I forked of VFPXDeployment, and testing my changes I do a lot of pushes to my repo to get the stuff back into the Thor folders. New option to add VFPX like docu, add options to fit it to  a non file git approach, working projects inherited, like that.  push - CFU; ad nauseam    
I test the results in my other projects - through CFU.   
I changed Bin2Text project to better integrate to Thor installed FoxBin2Prg.    

Stuff like that.

And there is stuff I like to understand how it is working. The working of Thor itself feels not well documented. There is stuff, in special in GoFish, I like changed, but I feel I need to understand it.

Never mind. The issue was on my list for a while, only now it was more often. But it looks like this comes to an end soon. Doing last polish to VFPXDeployment right now.

**Comment by Jimrnelson** [comment]
@lscheffler 

> I forked of VFPXDeployment, and testing my changes I do a lot of pushes to my repo to get the stuff back into the Thor folders. 

Oh, I see, but that seems inside out to me.  I do all my testing in the Thor folders in the first place.  Only when I am finally done with them (or I want to save my work along the way) do I move the changes into the repo.  That saves a lot of back and forth.

However, even the way you are working, instead of running CFU, all you need do is to manually unzip the file that is created for a project (into the right place).  That's all CFU is really doing for you.

> And there is stuff I like to understand how it is working. The working of Thor itself feels not well documented. There is stuff, in special in GoFish, I like changed, but I feel I need to understand it.

Yup, parts of Thor are well documented, parts not documented at all.  What is it you are interested for GF?  What do you want to change?   CFU delivers GoFish.APP, what else do you need to know?  

There's no reason to stay in the dark.  Please ask Thor CFU questions in this forum.

**Comment by lscheffler** [comment]
@Jimrnelson It's a bit a hen - egg problem, and one can not do all at once. To understand what goes to what folder in Thor is a bit a riddle. And coding inside the Thor folders I have done for a part of work, but then I do need to sync with the source, what is as well odd.
VFPXDeployment is circular, hard enough to have it in one structure.
So I decided to use push - CFU. :)

For GoFish the big task is to get rid of the 5 in GoFish5. In CFU, the menu and the app itself. GoFish will do. I have no idea how this works with old / new version. Or even what to change. And it must by synced somehow.

**Comment by Jimrnelson** [comment]
@lscheffler 

Lutz --

This is not as tricky as you imagine, but there are some sticky points.

[1]  Change every reference in GF that says "GoFish5" to be "GoFish".

[2]  Create a new updater for GoFish, changing references here:

![image](https://github.com/VFPX/Thor/assets/81179335/5ebf603d-635d-4e9e-ae93-b9db0c4be37c)

You should be able to try this locally first.  I can think of at least two sticky points:

[1]  Any menu (other than "Thor Tools", something I do not even use) that anybody has that references GoFish5 will still reference it, not the new version.

[2]  Any hot key assignment in GF that reference GoFish5 will still reference it, instead of the new 'GoFish'

For some people, this will be very annoying.  However, it IS possible for your GF installation program to correct these two problems.  We can discuss that when you get closer.

**Comment by lscheffler** [comment]
Yep that's  update prg. There is a bit of Thor menu that installs itself?

And then there are around  10 files *GoFish*something* I see on a Thor folder system without GoFish installed.

**Comment by Jimrnelson** [comment]
> Yep that's update prg. There is a bit of Thor menu that installs itself?

The menu "Thor Tools" is created by Thor from the contents of the tools it finds.  Change the contents of the tools, or add new tools, and the menu changes.
 
> And then there are around 10 files _GoFish_something*

Where are you seeing this?  

> I see on a Thor folder system without GoFish installed.

Where are you seeing this?

There most likely are various folders and files left around from previous installations of GF, but they don't matter.  They are no longer referenced, but no attempt has been made to erase them.

**Comment by lscheffler** [comment]
> I see on a Thor folder system without GoFish installed.
> Where are you seeing this? 

- "Install" Thor to a new empty folder
- run Thor. app from there an a plain IDE
- load at least PEM Editor (as suggested) and Thor Repository (should be suggested)
- use MS file search to search the folder for "GoFish"

![grafik](https://github.com/VFPX/Thor/assets/27769909/e21b022d-d82c-4028-bfb0-d597e41db5cf)

This looks like (installed temporary to *E:\SE\tmp\Thor_x*)    
![grafik](https://github.com/VFPX/Thor/assets/27769909/ddf666e5-f3da-458f-b01f-435188bbe76b)

**Comment by Jimrnelson** [comment]
Yup, there are a few other files in the system that have GoFish as part of their file name.  (Not even GoFish5, except for the updater).

These are not relevant or worrisome.  Ignore them.

**Comment by lscheffler** [comment]
Thank you!

So, now where are the places that define the menu entries? (for GoFish)

**Comment by Jimrnelson** [comment]
> So, now where are the places that define the menu entries? (for GoFish)

They are in the Thor_Tool_*  prgs created by GF as part of the installation process.  This is handled by GF already.

If you simply replace the GoFish5 references to GoFish in your GF project, the menus will change as well when you install.

**Comment by lscheffler** [comment]
I will rework GoFish to fit to VFPX Deployment and create the files while doing so. This will need some days.

**Comment by Jimrnelson** [comment]
@lscheffler 

We seem to have wandered off topic here.  Our last few messages have nothing to do with restoring projects after CFU.

**Comment by lscheffler** [comment]
ACK

**Comment by Jimrnelson** [comment]
This item, restoring projects after CFU, is now complete.

This was accomplished by creating plug-ins that are run before and after CFU.  Plug-ins are accessed here:

![image](https://github.com/VFPX/Thor/assets/81179335/1a8d0bca-ee6a-4901-b4e9-b5eaff956e73)

Which brings you to this screen where you can update the plug-ins if desired:

![image](https://github.com/VFPX/Thor/assets/81179335/3e25ce5f-cbe6-43a2-8eeb-c11a974050f1)

The default behavior for the plug-ins is to save the open projects (before) and re-open them (after).  You can add any other behaviors as desired my editing them from this screen.

**Comment by lscheffler** [comment]
Nice. Works, Thank you.

So the plugins are if I need to do something else then restoring the projects?

**Comment by Jimrnelson** [comment]
Something else or something more, as you wish.

If you start the process of editing one of them, you'll see a line which invokes the default behavior.

For my own use, I have modified the "After" plug-in to call my usual code for completely restoring my environment.

**Comment by lscheffler** [comment]
All my environment is in _SCREEN, PATH or menue - except the projects, so I'm fine. Thanks!

**Comment by lscheffler** [comment]
@Jimrnelson Hi Jim, just one question. Do you used my example straight away? In special, did you use `ADDPROPERTY(_SCREEN,"gaFiles(1,2)")`?

It looks like the property is gone after `Do RunThorPrg With n` and having a hit.

Possibly you can do Doug's trick, and name such stuff like gaThorFiles or aThorFiles or the like, so it does not interfere .... :)

I just copied the stuff from a general use, but now I have a bigger problem with the property gone. Thank you

**Comment by Jimrnelson** [comment]
@lscheffler 

Lutz --

I'm unclear here what you are reporting.  I took your code exactly as you provided it. which created a property in _Screen (before CFU) and then used and deleted it after CFU.

Are you saying that that property name is something you use distinct from this process?  It certainly would be easy enough for me to choose some other distinct property name (although I will not have time to make even this minor change for some time, as a work avalanche came late yesterday.)

**Comment by lscheffler** [comment]
@Jimrnelson This was the question. You used the code **exactly** like I provided it. Only it was just an example. :)   Looks like two of us have done the quick'n'dirty copy'n'paste. :D

Yes, **I** I use _SCREEN.gaFiles on several stuff and have several stuff stored there - while calling Thor.
I thought you would use it in the general form of the Thor Properties - no first "g" and with "Thor", just like "aThorFiles" or the like.

The problem will not raise the next 7 days with me, so no problem. I also can circumvent on this place, no big deal. Just bring it in line with Thor standards , if you find the time. This is not critical.

**Comment by Jimrnelson** [comment]
This item is being re-opened as the code to accomplish it causes other unanticipated problems, as identified by @lscheffler in issue #118

This feature will be disabled until those other problems have been resolved.

**Comment by Jimrnelson** [comment]
Re-opening

**Comment by lscheffler** [comment]
@Jimrnelson Is there any progress on this issue?  
It's still annoying to find an empty screen after starting VFP - normally it means I started VFP twice, so the projects are not re-instantiated at startup. And I like to proceed with some projects that require constant use of CFU ...

Thanks

**Comment by Jimrnelson** [comment]
@lscheffler 

I'll take another look.  It's been so long since this was closed it's worth another look,

**Comment by Jimrnelson** [comment]
@lscheffler 

With version 1.46.15, release today, you now have access to two new plug-ins, one of which runs just before CFU and one right after.

Create / modify them as needed.  (Restoring open projects, as I recall, but anything else of interest as well.)

![image](https://github.com/VFPX/Thor/assets/81179335/711843cd-04b1-4058-8eb5-4621f0b6a18a)


![image](https://github.com/VFPX/Thor/assets/81179335/8be3d50e-5d4c-416a-aa89-4dd549aaf618)

**Comment by lscheffler** [comment]
You say, without this plugin, CFU will leave the user without environment restored?

**Comment by Jimrnelson** [comment]
CFU has never done anything about restoring the user environment when it is done.    This would be a tricky thing to implement, as my guess would be that every user has their own interest in what they would want restored.

Thus, two new plug-ins for you allows you to customize as you see fit.

(BTW, in my own environment, I have recognized that the "After" plug-in works really nicely for me.  Kind of funny to have taken a dozen years for me to get to this point.)

**Comment by lscheffler** [comment]
> This would be a tricky thing to implement, as my guess would be that every user has their own interest in what they would want restored.

Simple. All what is touched and could be guessed. At least the build in stuff. Like SET, SYS, TOOLBAR - and the PJX. 

I do not see why you are so reluctant now. You tried before, only something else failed, couldn't you fix this?

**Comment by Jimrnelson** [comment]
Well, no, the list is infinite, as it does a CLOSE ALL and CLEAR ALL.

Thor has never tried to shoulder this responsibility and I see no reason to try for it to do so now.

In my personal case, the built-in stuff, as you say, would not be relevant.  

And over all these years nobody else has ever raised the topic.  I can't imagine why it would be worth the effort now.

**Comment by lscheffler** [comment]
> In **my personal case**, the built-in stuff, as you say, would not be relevant.

(－‸ლ)

Just to help you about **nobody else**, see https://github.com/VFPX/Thor/issues/96#issue-1716733517    The icon on bottom lists people who like to have it implemented.   

---
I do not understand why you first do the work in June, stopped it with #118 and now, after closing #118, you fight to do it not again. What the hell is today different to June?   
Why do you fixed #118 at all? It happened only if the projects where restored after CFU. Fixing #118 was a waste of time.

**Comment by Jimrnelson** [comment]
@lscheffler 

Please take the personal commentary out of this thread and all other threads.

I believe you are competent enough to keep technical discussions ... technical.

**Comment by lscheffler** [comment]
> And over all these years nobody else has ever raised the topic.

See https://github.com/VFPX/Thor/issues/96#issue-1716733517 The icon on bottom lists people who like to have it implemented.

**Comment by lscheffler** [comment]
> In my personal case, the built-in stuff, as you say, would not be relevant.

I beg your pardon to be frustrated on such statements. Nice for private projects. Thor is not private. In my not so humble view it is mandatory to do as good as possible.
You know nothing about users (better customers) environment, Thor is advertised is the Swiss knife of development, so possibly you should rethink to rethink this POV.

**Comment by lscheffler** [comment]
@Jimrnelson No need for the discussion https://github.com/VFPX/Thor/issues/219#issuecomment-1847921804.   
We are through this. See above

As I mentioned somewhere else, this is not my approach. I need some functionality, that you, at least as far as a I understand, reject, so I do the change to my fork and work from there. This is active for a while.   
Introducing the pjx storage is minor there.
I will not start playing with plug-ins. There is no need to learn
- how this interface works
- how to publish
- how to synch

I just can look up the merge where I had it running and start from there.

As expressed before, I publish a pull request if I consider it stable, and that's it.

Don't be sad, this makes a wonderful example for dealing with branches. Nothing is better to learn then real data. ;)

**Comment by Jimrnelson** [comment]
Implemented in Thor 1.47 - Released 2024-01-04 (#244)

**Comment by lscheffler** [comment]
(*.*)   
Thaanks
